### ISSUE TITLE
Makefile situation

### ISSUE THREAD

**Issue by digama0** [metamath/metamath-exe#23]
I have a makefile that says this, based on the readme:
```makefile
build:
	gcc m*.c -o metamath -Wall -pedantic -std=c99
```
Should I check this (or something like it) in? The current setup involves autoconf and generates a bewilderingly complex collection of scripts and files. @david-a-wheeler Do you know what's going on in `Makefile.am` and `confgure.ac`? Can't we make this project just have a simple makefile so that plain `make` works?

**Comment by wlammen** [comment]
A common Unix program compilation and installation has a configure script.  It allows the user for example to instruct the system where to install (globally, or just for the logged in user) the compiled program.  Autoconf helps the developer generating a script, that is valid on a wide range of platforms.
https://en.wikipedia.org/wiki/Configure_script

It is not that difficult to decipher the contents of configure.ac:
https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.70/autoconf.html#Autoconf-Language

**Comment by digama0** [comment]
I know what autoconf is. My point is that I don't see why our build setup does not need to be this complicated, and I'd like to know if there are extenuating circumstances making a basic makefile not just work.

**Comment by wlammen** [comment]
This might help:
https://stackoverflow.com/questions/719057/why-use-build-tools-like-autotools-when-we-can-just-write-our-own-makefiles

**Comment by digama0** [comment]
I've actually already read that, in an attempt to answer this question myself. I'm still not satisfied though, because we aren't using almost any of the features, it's literally a one line compiler invocation that is being replaced. What do you think about a project layout like https://github.com/kostrahb/Generic-C-Project ?

**Comment by digama0** [comment]
I took a more careful look at `configure.ac`. I see the following:

1. Initializes the version to metamath 0.114
2. Checks for `limits.h`, `stdlib.h`, `string.h`, `stdbool.h`
3. Uses `AC_TYPE_SIZE_T`, which apparently does "Define size_t to a suitable type, if standard headers do not define it."
4. Checks for `malloc`, `realloc`, `strchr`, `strcspn`, `strstr`
5. Test compiles a program to find out whether it is safe to enable `-Wall -Wextra`, and does so if possible
6. Test compiles a program with`-O3 -funroll-loops -finline-functions -fomit-frame-pointer` to find out if it is possible to add these flags
7. Test compiles a program using the `inline` keyword, and if so sets `-DINLINE=inline`, which has an effect on the source, which uses the `INLINE` macro where `inline` is expected

* Given that (1) is terribly out of date, it is clear to me that Norm never used this file in his workflow.
* For (2), I think we should just assert that all these files exist and not attempt anything fancy if they can't be found. 
* Same for (3): `size_t` is in the standard we're targeting.
* Same for (4): these functions should just exist and we aren't prepared to really fall back on anything else.
* For (5), just about every compiler that tries to have gcc-like flags supports these options even if they don't do anything. I don't think we should hesitate to use them and a test compilation is not necessary.
* For (6), I would be inlined to put these under some category like "extra optimizations" that can be chosen by the user, since they tend to increase compile time a lot. But I don't think we should worry about the flags not existing in the build script, at least until we have a concrete example of a build configuration that fails this.
* For (7), since we have upgraded the requirement to C99 anyway, the `inline` keyword is required to exist.

In short, there is nothing that this setup seems to accomplish beyond what a simple makefile can do, at least for this project.

**Comment by wlammen** [comment]
I agree that some tests in the configure.ac look outdated and their outcome can be assumed.  I think Norm has copied the file from another project and adapted it slightly (once).  Much as you want to copy the Makefile from the Generic-C project and adapt it to your needs.

When I download a metamath version from GitHub I routinely use the suggested autoreconf ... method to build the executable.  That was fine so far, and the build time never exceeded the time for auxiliary steps like setting up necessary directories.

So what are the respective pros and cons? With autotools you can rest assured you will have minimum hassle with users building metamath on whatever platform they use, including virtual machines, docker and the like. No matter in what direction metamath will develop.  The downside is that you need to learn the script language which might be a burden.  And there is some overhead on the technical side, because the flexibility of autotools appears unnecessary in some respects for a simple executable like metamath.

I miss the install option in the Makefile of Generic C.

**Comment by david-a-wheeler** [comment]
> Should I check this (or something like it) in? The current setup involves autoconf and generates a bewilderingly complex collection of scripts and files. @david-a-wheeler Do you know what's going on in Makefile.am and confgure.ac?

I know what's going on in Makefile.am and configure.ac; I wrote them. The autotools are old & quirky, but they're very well understood & automate a lot of things. I have a video tutorial about the autotools on Youtube if you find them bewildering: https://www.youtube.com/watch?v=4q_inV9M_us&list=PL1jSu6GGefBlGgdOeDpsKuCEy1ERNyCUZ

> Can't we make this project just have a simple makefile so that plain make works?

Is it *possible*? Yes. But "simple makefile" turns out to be not-so-simple in practice unless you want to be user-hostile and/or slow. Which is why tools like automake are helpful, they automate many things so we can spend time doing other things (like writing proofs!). Let's look at the proposed example:

> gcc m*.c -o metamath -Wall -pedantic -std=c99

If you *always* recompile *every* file and they are always `m*.c`, then recompiling `m*.c` every time would *run*. However, if you're editing metamath source code (instead of just doing a cold recompile) this is *far* slower than only compiling the code that needs to change. So while `m*.c` is fine in a special case (code compile) this is a *terrible* idea in general.

You can solve this by using make as it was originally intended: identify in the makefile every .c / .h / .o file, and manually identify their dependencies. This absolutely works as long as the code is never changed later. However, my experience is that over time the makefile always becomes wrong; it's just too easy to add an "include" and forget to modify the makefile, and since it works *most* of the time you end up with hair-pulling incomprehensible problems that may be months after the change (so it's hard to figure out "what went wrong").

You can solve for *that* by adding special compiler flags to generate makefile fragments and load them in, to automatically generate the dependency tree. Well, except those options aren't standard, and there are some weird special cases so you'll need to write some short scripts to modify them to work in all cases. Especially since different compilers & different compiler versions generate slightly different ones.

Also, that command assumes that you must use "gcc". That's popular, but not everyone uses it. You want `$(CC)` not `gcc` so the user can pick the compiler. Don't forget to add `$(CFLAGS)`.

Oh wait, the build command above doesn't *optimize* the resulting code, and there are some other useful flags for different compilers. Which flags? Well, that depends on the compiler. Better add more cases. That makefile sure isn't getting simple.

Wait, that build command doesn't install it anywhere. We'll need a "make install". That needs to support `DESTDIR` - does the makefile command support DESTDIR? Oh, and what about make uninstall? There are a lot of makefile conventions for expected functionality; [here's an introduction](https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html#Makefile-Conventions).

After you've done that... congrats, you just re-implemented a build system tool like automake, cmake, meson, etc.

It's quite possible to re-do all that in a straight makefile. I've even done it. But it's more work than you expect, and you quickly find yourself needing to use GNU make extensions & a *lot* more makefile code than you'd expect. I'd rather spend time doing other things.

Oh, and there's quite a bit to properly configuring code. Yes, a lot of stuff is now standard & doesn't need to be checked for. But don't assume that just because a standard says something exists that it actually works. It turns out that (gasp!!) some compilers don't fully implement some specs and some are *buggy* ; it's helpful to detect & work around those problems when practical. If you want many people to be able to pick up & use a tool, it's good to try to make the software "just work" on *many* platforms - even if it adds some scripts to make that happen. Also, why should the user need to spend time reading up how to configure the build so that the user needs to figure out that "CC" needs to be set & what to set CC to? Shouldn't the system do that automatically?

Oh, and does everyone run Linux? Probably not, and even if they did, there are many Linux distributions. What about MacOS? NetBSD? FreeBSD? OpenBSD? Do you really want users to manually have to configure their build? If not, do you want to write the configuration script to figure it out so it's automatic? Does everyone use GNU Make (no)?

I'm certain that the configure.ac / Makefile.am files need an update. Happy to do it. Norm liked to use his own idiosyncratic development approach, but since I suspect no one else does things the way he liked to do them, so we can focus on the current maintainers & users.

I picked the autotools because I and millions of other developers are familiar with them, and they automate a lot of things that you may not *realize* they automate until you discover they're missing.

We could  have a "straight makefile" but someone should put in the potentially dozens of hours needed to make that work to handle all common cases, not just the trivial ones. If we had to use a "straight makefile" then we'd probably soon have to give up & use GNU make extensions; portable make is a pain in the butt to use directly (but of course this loses portability).

We could use something else, e.g., meson is pretty popular. The autotools definitely show their age, and lot of people like some of the newer build systems. However, then someone will need to pick up the work to look at the alternatives & transition to them.

In any case, `gcc m*.c` is a *terrible* approach; it doesn't handle a vast number of cases. Note that the CI pipeline *does* use the autotools files, so they're getting used today.

**Comment by digama0** [comment]
@david-a-wheeler Re: `gcc m*.c` is a bad approach, I agree, I was just copying stuff from the readme, and I am no more an expert in writing makefiles than autoconf. But I understand that make is capable of tracking changed files just fine, and I just write some voodoo in there and promptly forget about it except for updating the list of files (which FYI we already have to do, there is such a list in `Makefile.am`). I'm not seriously proposing that exact makefile, but rather inviting us to write a proper makefile that does not have as many moving parts as the current setup. Something like the Generic-C makefile is closer to the real thing.

>You can solve this by using make as it was originally intended: identify in the makefile every .c / .h / .o file, and manually identify their dependencies. This absolutely works as long as the code is never changed later. However, my experience is that over time the makefile always becomes wrong; it's just too easy to add an "include" and forget to modify the makefile, and since it works most of the time you end up with hair-pulling incomprehensible problems that may be months after the change (so it's hard to figure out "what went wrong").

I don't see how the current setup solves this problem at all. There is an explicit list of c and h files in `Makefile.am`, and if this gets out of sync then you will have these exact issues. If anything, Norm's `m*.c` solves this problem since it just picks up everything in the directory.

> Also, that command assumes that you must use "gcc". That's popular, but not everyone uses it. You want $(CC) not gcc so the user can pick the compiler. Don't forget to add $(CFLAGS).

I seriously don't understand the motivation behind this. We can have a `CC=gcc` instruction in the makefile if you want to make it easy to swap the compiler for something else but I really don't see the point. (The Generic C makefile uses `cc` for this.)

> Oh wait, the build command above doesn't optimize the resulting code, and there are some other useful flags for different compilers. Which flags? Well, that depends on the compiler. Better add more cases. That makefile sure isn't getting simple.

This sounds like a problem created by trying to target many compilers. We can support *multiple* compiler invocations, but this kind of "maybe it works if you just change the compiler invocation" support isn't support at all and I dislike the philosophy behind it.

Many compilers have similar flags, so a `$(CC)` compiler invocation and testing in CI for the common compilers is sufficient.

> Wait, that build command doesn't install it anywhere. We'll need a "make install". That needs to support DESTDIR - does the makefile command support DESTDIR? Oh, and what about make uninstall? There are a lot of makefile conventions for expected functionality; here's an introduction.

This sounds like something I can copy and paste from a makefile intro somewhere. The install procedure for metamath is very straightforward since it's just one binary file. It should not require changes ever.

> Oh, and there's quite a bit to properly configuring code. Yes, a lot of stuff is now standard & doesn't need to be checked for. But don't assume that just because a standard says something exists that it actually works. It turns out that (gasp!!) some compilers don't fully implement some specs and some are buggy ; it's helpful to detect & work around those problems when practical.

Can we just not? I really don't think that's a good way to write C code. Luckily the majority of problems of this nature that we have to deal with at this point only exist because of backward compatibility and can be fixed by upgrading the targeted version. Certainly this kind of code does not square well with the aspirations in the other thread for CompCert verified code stuff.

> If you want many people to be able to pick up & use a tool, it's good to try to make the software "just work" on many platforms - even if it adds some scripts to make that happen. 

I want every such workaround to have testing in CI for it. Configuration fixes like this have a tendency to rot, and they contribute significantly to the overall maintenance burden.

> Also, why should the user need to spend time reading up how to configure the build so that the user needs to figure out that "CC" needs to be set & what to set CC to? Shouldn't the system do that automatically?

They absolutely should not have to configure the build. I want `make` to Just Work. I will note that although `autoconf` is supposed to solve this problem, it does not Just Work on windows so I'm less than impressed with its ability to deliver on its goals.

> Oh, and does everyone run Linux? Probably not, and even if they did, there are many Linux distributions. What about MacOS? NetBSD? FreeBSD? OpenBSD? Do you really want users to manually have to configure their build? If not, do you want to write the configuration script to figure it out so it's automatic? Does everyone use GNU Make (no)?

Are these being tested? Because if not then my default assumption is that they don't work.

> We could have a "straight makefile" but someone should put in the potentially dozens of hours needed to make that work to handle all common cases, not just the trivial ones. If we had to use a "straight makefile" then we'd probably soon have to give up & use GNU make extensions; portable make is a pain in the butt to use directly (but of course this loses portability).

It boggles my mind that anyone can spend so much time on build maintenance. There is *nothing* special about this project compared to trivial (multi-file) C programs you can find anywhere on the internet, so we should have a Makefile that reflects the 10 minutes of copy paste work needed to get something decent. Attempting to support all compilers on all systems is a fool's errand, but making something that works on a finite enumerated set of compilers and systems is not that hard.

**Comment by david-a-wheeler** [comment]
>  I'm not seriously proposing that exact makefile, but rather inviting us to write a proper makefile that does not have as many moving parts as the current setup. Something like the Generic-C makefile is closer to the real thing.

I'm not morally opposed to moving from automake, it's quite clunky. But if we do, I would want the result to have its main end-user functionality, e.g., so we can easily and *correctly* recompile *only* what needs to be recompiled, easily install using DESTDIR, easily uninstall, etc.

The obvious solutions are a simple makefile using GNU make extensions, CMake, or meson.

>> You can solve this by using make as it was originally intended: identify in the makefile every .c / .h / .o file, and manually identify their dependencies. This absolutely works as long as the code is never changed later. However, my experience is that over time the makefile always becomes wrong; it's just too easy to add an "include" and forget to modify the makefile, and since it works most of the time you end up with hair-pulling incomprehensible problems that may be months after the change (so it's hard to figure out "what went wrong").

> I don't see how the current setup solves this problem at all. There is an explicit list of c and h files in Makefile.am, and if this gets out of sync then you will have these exact issues.

It solves the transitive dependency problem. You're conflating 2 different issues, which is completely understandable because they look similar at first. If you forget to list a .c file in the Makefile.am file, the program won't compile at all, which is easy to detect. That's not the problem that automake is trying to solve.

The problem is incorrectly specified transitive .h dependencies.  Let's say that a.c depends on a.h, and a.h includes b.h. The obvious but often wrong solution in make is `a.c: a.h` ... forgetting that a.b transitively depends on b.h, but a.h is not *generated* from b.h. Then if b.h is changed, the file `a.c` is not recompiled, leading to extremely mysterious errors.

> If anything, Norm's m*.c solves this problem since it just picks up everything in the directory.

That solves the cold compile problem, but not the "compile after a source file change" problem. Recompiling everything is terrible for an edit-compile-test cycle.

>> Also, that command assumes that you must use "gcc". That's popular, but not everyone uses it. You want $(CC) not gcc so the user can pick the compiler. Don't forget to add $(CFLAGS).
> I seriously don't understand the motivation behind this. We can have a CC=gcc instruction in the makefile if you want to make it easy to swap the compiler for something else but I really don't see the point. (The Generic C makefile uses cc for this.)

While gcc is common, many systems don't have gcc but have a working C compiler. The point of "./configure" is to automatically figure out "something that works" instead of requiring the end-user to figure out how to make it work. We can document "make CC=your_compiler" - that. simplifies the build system by making it harder on some end-users.

> This sounds like a problem created by trying to target many compilers. We can support multiple compiler invocations, but this kind of "maybe it works if you just change the compiler invocation" support isn't support at all and I dislike the philosophy behind it.

It *is* supported. That's precisely what automake/autoconf do, we *already* support multiple compiler invocations.

We don't have CI tests for it, but that's easily fixed. Even if we stop using automake, we should still test it with multiple compilers. Adding more CI tests is something I was nagging Norm to do anyway.

>> Wait, that build command doesn't install it anywhere. We'll need a "make install". That needs to support DESTDIR - does the makefile command support DESTDIR? Oh, and what about make uninstall? There are a lot of makefile conventions for expected functionality; here's an introduction.

> This sounds like something I can copy and paste from a makefile intro somewhere. The install procedure for metamath is very straightforward since it's just one binary file. It should not require changes ever.

Yes, it's relatively straightfoward because it's just one binary file. I can even work out that Makefile (sample below). It's just more complicated than you might expect.

> I want every such workaround to have testing in CI for it. Configuration fixes like this have a tendency to rot, and they contribute significantly to the overall maintenance burden.

We're on the same page for CI testing.

> Are these (other systems) being tested? Because if not then my default assumption is that they don't work.

We could & should add such tests.  Metamath doesn't use any complex systems & is pretty warning-free, so the odds are pretty good.

> It boggles my mind that anyone can spend so much time on build maintenance. There is nothing special about this project compared to trivial (multi-file) C programs you can find anywhere on the internet, so we should have a Makefile that reflects the 10 minutes of copy paste work needed to get something decent. Attempting to support all compilers on all systems is a fool's errand, but making something that works on a finite enumerated set of compilers and systems is not that hard.

"All" is rediculous, agreed, but systems vary more than you might think. It's easy to create a Makefile that doesn't work for a lot of people. "It works for me" doesn't mean it works for anyone else.

I have a lot of experience creating makefiles, so I can gen up something reasonable. But it's more than you might think.

If you are willing to switch to GNU make (instead of pure POSIX standard make), and use just gcc or clang, you can auto-generate a lot of dependencies this way: http://www.microhowto.info/howto/automatically_generate_makefile_dependencies.html

So you end up with a makefile that looks like this:

~~~~
CPPFLAGS += -MD -MP

BINDIR = /usr/local/bin
INSTALL = install
EXE = # Set to ".exe" on Windows
RM = rm -f

SRC = $(wildcard *.c)

metamath$(EXE): $(SRC:%.c=%.o)
	$(LD) -o $@ $^

install:
	$(INSTALL) metamath$(EXE) $(DESTDIR)$(BINDIR)

uninstall:
	$(RM) $(DESTDIR)$(BINDIR)/metamath$(EXE)

-include $(SRC:%.c=%.d)
~~~~

**Comment by david-a-wheeler** [comment]
As with all things, there's a trade-off. The much-simpler makefile means that there's no auto-detection of their local configurations, so end-users end up having to figure out how to make it work in many cases, making it harder on them. If we think that's an acceptable trade-off, then that's a way to proceed.

**Comment by david-a-wheeler** [comment]
> It boggles my mind that anyone can spend so much time on build maintenance. There is nothing special about this project compared to trivial (multi-file) C programs you can find anywhere on the internet, so we should have a Makefile that reflects the 10 minutes of copy paste work needed to get something decent.

It's really common to use something *OTHER* than trivial POSIX-standard makefiles for C programs once the sources are distributed. That's why GNU extensions, automake, Cmake, meson, & other tools exist. You start needing extensions (GNU make), external generators (automake/Cmake), or a completely new tool (meson) unless you want mysterious errors.

**Comment by david-a-wheeler** [comment]
Oh, to answer your original question: https://github.com/kostrahb/Generic-C-Project suffers from a common defect, it assumes that .h file dependencies never change. As long as they never change it works fine. If they do, you end up with mysterious hard-to-debug errors. Which is why you want to use the "include" magic. Also it uses "cc" instead of the more correct & user-controllable $(CC). It, just like my script above, requires GNU make. It's missing some other targets like "clean" but of course we could add those.

**Comment by eli-schwartz** [comment]
It actually does

```
# Generate dependencies
%.d: %.c
	$(CC) -MM -MG -MF $@ $<
```

Properly using $(CC), improperly generating depfiles without -MD as a side effect of normal compilation -- instead they are generated manually as an additional step, essentially meaning it is an alias for https://en.wikipedia.org/wiki/Makedepend.

It also uses recursive Make!

And then generates .o object files via a submake, using `$(CC)`, then links them together in the parent make using `cc`. So you can end up compiling with one compiler and linking with a different compiler.

I do not want to think about how many edge cases this will cause. Please stick to autotools. :p As a developer of a build system (meson) who also thinks plain Makefiles can be okay if you are careful and thorough (the one that you just posted is a good basic example)... this example project thing hurts my soul. *Everyone* loses out if we allow this sort of thing to become common.

**Comment by david-a-wheeler** [comment]
@eli-schwartz : I see. The *top* level Makefile doesn't generate dependencies, but it also includes a src/Makefile that *does* do that. Well, if you split up functionality into enough different files, it's harder to maintain but at least each file looks simpler :-). Recursive make is almost always a bad idea, that's a great way to implement GIGO (garbage in, garbage out).

**Comment by digama0** [comment]
It sounds like most others are in favor of keeping the autotools build process despite its complexity. In that case, I would suggest the following:

* Simplify `configure.ac` to remove the compile tests
* Put sources in a `src` folder, and ensure that autoconf puts all generated files in the `build` folder
* Add a `.gitignore` that ignores the `build` folder
* Add a mode for debug builds / "fast" builds using autoconf

The current autoconf approach is way slower than the dumb `gcc *.c` makefile - even if we recompile everything from scratch it still only takes about 1.5s to compile on my machine. Compiling `gcc *.c -O3 -funroll-loops -finline-functions -fomit-frame-pointer` is about 17s, while `autoreconf -i && ./configure && make` is 2.5s + 2.2s + 23.9s.

**Comment by eli-schwartz** [comment]
Autoconf inevitably introduces quite a bit of slowdown during the first time configuring, due to running the configure script. Shell scripts aren't particularly fast, and the more you fork and run external programs (which is how nearly everything has to be done) the slower it gets.

You could in theory switch to meson (or cmake) and do configure checks and Makefile / build.ninja generation in-process. Particularly, using ninja instead of Make can be extremely fast, as ninja prioritizes blazing-fast parsing and rule execution in return for not supporting any of the templating, pattern and macro features which Make (and especially GNU make) does.

This is of course not quite as portable.
Ninja does build on a number of common platforms, but it's probably not installed by default which could be a bit inconvenient.
Meson requires python, and is generally faster than autotools but still required some time to warm up the python VM... though its c99 reimplementation https://sr.ht/~lattis/muon (WIP but handles most of the common parts of meson's build language) solves both those issues.

Cmake is not a solution to make anything *less* complex. :p

**Comment by eli-schwartz** [comment]
A very easy speed boost would be to simply drop every "AC_CHECK anything", unless you actually act on the results by doing ifdefs on some `HAVE_` macro as output by the checks. And ask whether any given `HAVE_` ifdef is important enough compared to just not supporting broken systems.

**Comment by digama0** [comment]
Could someone check this autoconf script for me? This is simplified from a combination of the current `configure.ac` and the output of `autoscan`.
```autoconf
#                                               -*- Autoconf -*-
# Process this file with autoconf to produce a configure script.

AC_PREREQ([2.65])
AC_INIT([metamath], [0.114], [nm.NOSPAM@alum.mit.edu])
AC_CONFIG_SRCDIR([metamath.c])
AC_CONFIG_HEADERS([config.h])

AM_INIT_AUTOMAKE([foreign])
AC_CONFIG_FILES([Makefile])

# Checks for programs.
AC_PROG_CC
AC_PROG_INSTALL

# Use --std=c99
AC_PROG_CC_C99

# Use inline if possible
AC_C_INLINE

# Checks for header files.
# AC_CHECK_HEADERS([limits.h stdlib.h string.h])
# AC_HEADER_STDBOOL

# Checks for typedefs, structures, and compiler characteristics.
# AC_TYPE_SIZE_T

# Checks for library functions.
# AC_FUNC_MALLOC
# AC_FUNC_REALLOC
# AC_CHECK_FUNCS([strchr strcspn strpbrk strstr])

CFLAGS="$CFLAGS -Wall -Wextra"

echo "CFLAGS=$CFLAGS"

AC_OUTPUT
```

I haven't figured out how to set up the debug/release thing yet. Adding `-O3` unconditionally is not correct, since this precludes a debug build, but I don't want people to always have to specify a bunch of C flags in the `./configure` invocation either.

**Comment by eli-schwartz** [comment]
Meson and cmake both have a built-in option "build type" which can be set to "debug" or "release" or "debugoptimized" or "plain" or things of that I'll, and automatically set up -g and -O flags based on that.

For autotools you can always do `AC_ARG_ENABLE([debug], ...)` and manually define the flags that should be set when --enable-debug is specified as a configure arg.

**Comment by digama0** [comment]
I'm going to consider this issue resolved by the build script overhaul in #54 .
